uberfire-layout-editor/uberfire-layout-editor-client/src/main/java/org/uberfire/ext/layout/editor/client/components/LayoutComponentView.java

```diff
@@ -44,112 +50,159 @@
+        RenderingContext renderingContext = new RenderingContext(layoutComponent, fluidContainer);
+        IsWidget previewWidget = type.getPreviewWidget(renderingContext);
+
+        Column buttonColumn = new Column(12);
+        buttonColumn.getElement().getStyle().setProperty("textAlign", "left");
+        if (previewWidget != null) {
+            buttonColumn.add(previewWidget);
+        }
         return buttonColumn;
     }
 
-    private Column generateButtonColumn() {
-        Column buttonColumn = new Column( 4 );
-        buttonColumn.getElement().getStyle().setProperty( "textAlign", "right" );
-        if ( type.hasConfigureModal() ) {
-            buttonColumn.add( generateConfigureButton() );
+    private Column generateHeaderRow() {
+        final Column header = new Column(12);
+        header.getElement().getStyle().setProperty( "textAlign", "right" );
+        if ( type instanceof HasConfiguration ) {
+            header.add(generateConfigureButton());
         }
-        buttonColumn.add( generateRemoveButton() );
-        return buttonColumn;
+        header.add(generateRemoveButton());
+        return header;
     }
 
     private Button generateConfigureButton() {
         Button remove = GWT.create( Button.class );
         remove.setSize( ButtonSize.MINI );
-        remove.setType( ButtonType.PRIMARY );
-        remove.setIcon( IconType.EDIT );
-        remove.getElement().getStyle().setProperty( "marginRight", "3px" );
+        remove.setType(ButtonType.PRIMARY);
+        remove.setIcon(IconType.EDIT);
+        remove.getElement().getStyle().setProperty("marginRight", "3px");
         remove.addClickHandler( new ClickHandler() {
             public void onClick( ClickEvent event ) {
-                showConfigureModal();
+                showConfigurationScreen();
             }
         } );
         return remove;
     }
 
-    private void showConfigureModal() {
-        final Modal configureModal = type.getConfigureModal( layoutComponentWidgetUI );
-        configureModal.show();
+    private void showConfigurationScreen() {
+        LayoutEditorWidget layoutEditorWidget = getLayoutEditorWidget();
+        LayoutComponent layoutComponent = layoutEditorWidget.getLayoutComponent(componentEditorWidget);
+
+        if (type instanceof HasModalConfiguration) {
+            ModalConfigurationContext ctx = new ModalConfigurationContext(layoutComponent, fluidContainer, this);
+            Modal configModal = ((HasModalConfiguration) type).getConfigurationModal(ctx);
+            configModal.show();
+        }
+        else if (type instanceof HasPanelConfiguration) {
+            PanelConfigurationContext ctx = new PanelConfigurationContext(layoutComponent, fluidContainer, this);
+            Panel configPanel = ((HasPanelConfiguration) type).getConfigurationPanel(ctx);
+            componentEditorWidget.getWidget().clear();
+            componentEditorWidget.getWidget().add(configPanel);
+        }
```

Comment:
As we've talked I personally don't like this approach with "instanceof" operator. I think here it would be better use a Enum as a type, as ask the type to showConfigurationScreen (passing necessary parameters as well)